### PR TITLE
[CFE/minmax-embedder] Introduce minmax-embedder library

### PR DESCRIPTION
**Pull request by glistening**
It introduces minmax-embedder, which embedds minmax recording in hdf5 to circle.

ONE-DCO-1.0-Signed-off-by: Sanggyu Lee <sg5.lee@samsung.com>

Draft: #11649

- negative test run
```
39: [==========] Running 1 test from 1 test suite.
39: [----------] Global test environment set-up.
39: [----------] 1 test from MinMaxEmbedderTest
39: [ RUN      ] MinMaxEmbedderTest.invalid_input_NEG
39: Failed to open file: not_existing
39: [       OK ] MinMaxEmbedderTest.invalid_input_NEG (0 ms)
39: [----------] 1 test from MinMaxEmbedderTest (0 ms total)
39: 
39: [----------] Global test environment tear-down
39: [==========] 1 test from 1 test suite ran. (0 ms total)
39: [  PASSED  ] 1 test.
1/1 Test #39: minmax_embedder_test .............   Passed    0.01 sec

The following tests passed:
	minmax_embedder_test

100% tests passed, 0 tests failed out of 1
```
- value test (#11699) will be added in forthcoming PRs.

### CODE REVIEW

**Comment by jinevening** [comment]
@nnfw-bot test nncc-release

File: compiler/minmax-embedder/src/Embedder.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  {
+    auto graph = module->graph(idx);
+
+    /* read subgraph inputs */
+    const auto input_nodes = loco::input_nodes(graph);
+    const auto n_inputs = input_nodes.size();
+    for (size_t input_idx = 0; input_idx < n_inputs; ++input_idx)
+    {
+      const auto *circle_input = loco::must_cast<const luci::CircleInput *>(input_nodes[input_idx]);
+      if (not circle_input || circle_input->index() != input_idx)
```
Comment (jinevening) [inline compiler/minmax-embedder/src/Embedder.cpp:98]: `must_cast` ensures that circle_input is not nullptr.

```suggestion
      if (circle_input->index() != input_idx)
```

File: compiler/minmax-embedder/src/h5/Reader.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+#include <sstream>
+#include <stdexcept>
+
+namespace minmax_embedder
+{
+namespace h5
+{
+static const char *h5_value_grpname = "value";
+
+bool exists(hid_t id, const char *path) { return H5Lexists(id, path, H5P_DEFAULT) > 0; }
```
Comment (jinevening) [inline compiler/minmax-embedder/src/h5/Reader.cpp:29]: These are used only in this file. Could you move them to anonymous namespace?

File: compiler/minmax-embedder/src/Embedder.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        throw std::runtime_error("Input order in minmax recording does not match to circle");
+
+      auto minmax = mmr.read_input(0, idx, input_idx);
```
[the remark this answers is not in the document]
Comment (glistening) [reply compiler/minmax-embedder/src/Embedder.cpp:101]: Yes. The interfave is in general form. Multiple models are not supported in minmax quantizer using onert. I will add explicit check code.

File: compiler/minmax-embedder/include/minmax-embedder/Embedder.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+#include <string>
+
+namespace minmax_embedder
+{
+
+struct EmbedderOptions
+{
+  float min_percentile;
+  float max_percentile;
```
[the remark this answers is not in the document]
Comment (glistening) [reply compiler/minmax-embedder/include/minmax-embedder/Embedder.h:28]: I don't set default value to keep this structure POD. I wonder SE tool always complains when structure definition does not have initial value?

[the remark this answers is not in the document]
Comment (glistening) [reply compiler/minmax-embedder/include/minmax-embedder/Embedder.h:28]: https://github.com/Samsung/ONE/blob/d5a631f08e950e876d4eb7445a4dd24d619216e2/runtime/onert/api/include/nnfw.h#L175-L186

`onert` uses `struct` without initializer.

[the remark this answers is not in the document]
Comment (glistening) [reply compiler/minmax-embedder/include/minmax-embedder/Embedder.h]: I am not sure SE tool really complain about `struct` w/o initial value. But I will follow the guide of Compile FE.

**Comment by glistening** [comment]
@jinevening PTAL. I've updated as you commented.

File: compiler/minmax-embedder/requires.cmake [hunk 5]
```diff
@@ -0,0 +1 @@
+require("loco")
```
Comment (hseok-oh) [inline compiler/minmax-embedder/requires.cmake:1]: Is it requires "loco" only?

```suggestion
require("safemain")
require("loco")
require("luci")
require("arser")
require("vconone")
```

Comment (glistening) [reply compiler/minmax-embedder/requires.cmake:1]: `arser`, `vconone`, `safemain` is not required. But `luci` is required. I will add it.

**Review by jinevening — DISMISSED** [review]
LGTM

File: compiler/minmax-embedder/requires.cmake (hunk 5 above)
Comment (hseok-oh) [reply compiler/minmax-embedder/requires.cmake:1]: Then why `arser`, `vconone`, and `safemain` is linked to `minmax_embedder` in cmake's `target_link_libraries` ?

Comment (glistening) [reply compiler/minmax-embedder/requires.cmake:1]: @hseok-oh You're right. I removed them from library dependencies. They should be linked for `driver`, which will be the next PR.

**Review by jinevening — APPROVED** [review]
LGTM

**Review by hseok-oh — APPROVED** [review]
LGTM

**Pull request merged**